shop/src/utils/generateVariants.js

```diff
@@ -1,25 +1,42 @@
-import flatMap from 'lodash/flatMap'
+const combinationsOfElementAndArray = (element, arr) => {
+  let accumulator = []
+  if (arr.length === 0) {
+    accumulator = [...accumulator, [element, arr[0]]]
+  }
+  for (let i = 0; i < arr.length; i++) {
+    if (Array.isArray(arr[i])) {
+      accumulator = [...accumulator, [element, ...arr[i]]]
+    } else {
+      accumulator = [...accumulator, [element, arr[i]]]
+    }
+  }
+  return accumulator
+}
 
-export const getAllCombinations = (array1, ...arrays) => {
-  const combinationsOf2Array = (a, b) =>
-    a.reduce(
-      (combinations, v1) => [
-        ...combinations,
-        ...b.map((v2) => flatMap([v1, v2]))
-      ],
-      []
-    )
+const combinationsOfTwoArrays = (array1, array2) => {
+  let combinations = []
+  if (array1.length === 0) {
+    combinations = [
+      ...combinations,
+      ...combinationsOfElementAndArray(array1[0], array2)
+    ]
+  }
+  for (let j = 0; j < array1.length; j++) {
+    combinations = [
+      ...combinations,
+      ...combinationsOfElementAndArray(array1[j], array2)
+    ]
+  }
+  return combinations
+}
```

Comment:
nit: In the loop if you change the constraint from `j < array1.length` to `j <= array1.length` you won't need the `if` condition above

Reply:
Touché

Reply:
Following up on this suggestion, I find that if array1.length is say, 1, the loop will execute `combinationsOfElementAndArray` twice - once with the `array1[0]`, and then again with `array[1]` (undefined)

Could you try running getAllCombinations(["a"], []) with your method? 
Preserving the `if` condition results in:
`getAllCombinations(["a"], []) = [["a", undefined]]`